### PR TITLE
Make Stack IR a canonical representation

### PR DESCRIPTION
**Pull request by tlively**
Rather than a temporary side data structure. This means that it is now
an error to run a normal pass in the presence of Stack IR because the
normal IR may be out of date and this can no longer be solved by
throwing away the Stack IR. The only way to transition back from
having Stack IR right now is to perform a round trip, so the fuzzer
and various tools are updated to add these round trip transitions
where necessary to avoid an error.

This change is a prerequisite for being able to lower tuples away when
generating Stack IR because lowering tuples requires changing a
Function's locals, which invalidates the normal IR.

### CODE REVIEW

File: test/debugInfo.fromasm [hunk 1]
```diff
@@ -2,48 +2,39 @@
  (type $i32_=>_i32 (func (param i32) (result i32)))
  (type $i32_i32_=>_i32 (func (param i32 i32) (result i32)))
  (type $none_=>_none (func))
- (import "env" "memory" (memory $memory 256 256))
- (data (global.get $__memory_base) "debugInfo.asm.js")
- (import "env" "__memory_base" (global $__memory_base i32))
+ (import "env" "memory" (memory $1 256 256))
+ (data (global.get $gimport$0) "debugInfo.asm.js")
+ (import "env" "__memory_base" (global $gimport$0 i32))
```
Comment (tlively) [inline test/debugInfo.fromasm:7]: I haven't been able to get these names to survive the round trip. @kripken, is this expected to work or is Binaryen incapable of emitting some names?

File: scripts/fuzz_opt.py [hunk 2]
```diff
@@ -618,7 +618,7 @@ def test_one(random_input, opts, allow_autoreduce):
                 # gets commands from the handler, for a given set of optimizations. this is all the commands
                 # needed to run the testing that that handler wants to do.
                 def get_commands(opts):
-                    return testcase_handler.get_commands(wasm='a.wasm', opts=opts + FUZZ_OPTS + FEATURE_OPTS, random_seed=random_seed)
+                    return testcase_handler.get_commands(wasm='a.wasm', opts=FUZZ_OPTS + opts + FEATURE_OPTS, random_seed=random_seed)
```
Comment (tlively) [inline scripts/fuzz_opt.py:621]: `FUZZ_OPTS` has to come before `opts` because `FUZZ_OPTS` might contain a JS legalization pass that cannot operate on Stack IR.

File: test/debugInfo.fromasm (hunk 1 above)
Comment (tlively) [reply test/debugInfo.fromasm:7]: Ah, actually this is causing the current test failures. wasm2js only adds `__wasm_fetch_high_bits` if it sees a global named `i64toi32_i32$HIGH_BITS`, but global names are being blown away by the round trip pass.

Comment (tlively) [reply test/debugInfo.fromasm:7]: After discussing with @kripken, this was fixed by not running stack IR passes in wasm2js.

File: README.md [hunk 3]
```diff
@@ -78,10 +78,10 @@ There are a few differences between Binaryen IR and the WebAssembly language:
    * WebAssembly limits block/if/loop types to none and the concrete value types
      (i32, i64, f32, f64). Binaryen IR has an unreachable type, and it allows
      block/if/loop to take it, allowing [local transforms that don't need to
-     know the global context][unreachable]. As a result, Binaryen's default
-     text output is not necessarily valid wasm text. (To get valid wasm text,
-     you can do `--generate-stack-ir --print-stack-ir`, which prints Stack IR,
-     this is guaranteed to be valid for wasm parsers.)
+     know the global context][unreachable]. As a result, Binaryen's default text
+     output is not necessarily valid wasm text. (To get valid wasm text, you can
+     do `--generate-stack-ir --print`, which prints Stack IR, this is guaranteed
```
[the remark this answers is not in the document]
Comment (tlively) [reply README.md:83]: `--print` only prints the module to stdout. If test modules are being printed twice, there's probably also logic in the test runner for printing them a second time. I'll look into that.

Correct, there is no stack IR parser. I had to add round trips to all tests that ended up with stack IR but needed to emit parsable output.

File: scripts/fuzz_opt.py [hunk 4]
```diff
@@ -603,7 +603,7 @@ def test_one(random_input, opts, given_wasm):
     print('pre wasm size:', wasm_size)
 
     # create a second wasm for handlers that want to look at pairs.
-    generate_command = [in_bin('wasm-opt'), 'a.wasm', '-o', 'b.wasm'] + opts + FUZZ_OPTS + FEATURE_OPTS
+    generate_command = [in_bin('wasm-opt'), 'a.wasm', '-o', 'b.wasm'] + FUZZ_OPTS + opts + FEATURE_OPTS
```
[the remark this answers is not in the document]
Comment (tlively) [reply scripts/fuzz_opt.py:606]: `FUZZ_OPTS` might contain a pass to do JS legalization, which can only run on normal IR rather than stack IR. Running `FUZZ_OPTS` after `opts` crashes when that is the case and `opts` ends with Stack IR. Changing their order was a simple workaround.

File: scripts/fuzz_opt.py [hunk 5]
```diff
@@ -655,7 +655,16 @@ def write_commands(commands, filename):
 
 opt_choices = [
     [],
-    ['-O1'], ['-O2'], ['-O3'], ['-O4'], ['-Os'], ['-Oz'],
+    # Even though -O0 and -O1 by themselves do not generate stack IR, the
+    # optimization level may be overridden by later optimization flags so we
+    # have to be defensive and add round trips for them as well.
```
[the remark this answers is not in the document]
Comment (tlively) [reply scripts/fuzz_opt.py:660]: The way optimization option parsing works, each instance of an `-O...` flag adds a marker to the pass sequence where a set of default optimization passes should be inserted and also overrides the current shrink and optimization levels. After all the options are parsed, the markers are replaced with the optimization passes for the last set optimization and shrink levels, so `-O1 -O3` is identical to `-03 -03`.

All other passes except `--generate-stack-ir` are safe in the presence of these overrides because they do not end with stack IR under any optimization or shrink levels.

File: src/pass.h [hunk 6]
```diff
@@ -297,6 +291,14 @@ class Pass {
   // out any Stack IR - it would need to be regenerated and optimized.
   virtual bool modifiesBinaryenIR() { return true; }
 
+  // Whether this pass operates on StackIR. Once StackIR is generated, the
+  // normal Binaryen IR is no longer valid, so it is an error to run a
+  // non-StackIR pass after that point.
+  virtual bool acceptsStackIR() { return false; }
+
+  // Whether this pass operates on nested IR.
+  virtual bool acceptsNestedIR() { return true; }
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/pass.h:300]: I'm not sure. I think for this PR it probably makes more sense to stick to the `StackIR`/`BinaryenIR` naming. In the future if Binaryen IR ends up getting a flat form to replace Stack IR we can figure out a better naming scheme.

File: src/passes/pass.cpp [hunk 7]
```diff
@@ -651,61 +656,55 @@ void PassRunner::doAdd(std::unique_ptr<Pass> pass) {
...
 
-  // In the creator we can scan the state of the module and function before the
-  // pass runs.
-  AfterEffectFunctionChecker(Function* func) : func(func), name(func->name) {
-    beganWithStackIR = func->stackIR != nullptr;
-    if (beganWithStackIR) {
+  AfterEffectFunctionChecker(Pass* pass, Function* func)
+    : pass(pass), func(func), name(func->name) {
+    // TODO: Check the Stack IR did not change if the pass does not accept it
+    checkHash = !pass->modifiesBinaryenIR() || !pass->acceptsNestedIR();
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/pass.cpp:674]: No, for example an analysis pass could operate over Stack IR without changing it. `modifiesBinaryenIR` should probably be renamed to `modifiesIR`.

File: src/passes/pass.cpp [hunk 8]
```diff
@@ -651,61 +656,55 @@ void PassRunner::doAdd(std::unique_ptr<Pass> pass) {
...
   HashType originalFunctionHash;
 
-  // In the creator we can scan the state of the module and function before the
-  // pass runs.
-  AfterEffectFunctionChecker(Function* func) : func(func), name(func->name) {
-    beganWithStackIR = func->stackIR != nullptr;
-    if (beganWithStackIR) {
+  AfterEffectFunctionChecker(Pass* pass, Function* func)
+    : pass(pass), func(func), name(func->name) {
+    // TODO: Check the Stack IR did not change if the pass does not accept it
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/pass.cpp:673]: No, we only check that the Binaryen IR did not change. We never check that the Stack IR doesn't change.

File: src/passes/passes.h [hunk 9]
```diff
@@ -83,7 +83,7 @@ Pass* createPrinterPass();
 Pass* createPrintCallGraphPass();
 Pass* createPrintFeaturesPass();
 Pass* createPrintFunctionMapPass();
-Pass* createPrintStackIRPass();
+Pass* createReloadModulePass();
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/passes.h:86]: Oops, that was left over from previous work before I realized that `--roundtrip` already existed.

File: src/wasm/wasm-validator.cpp [hunk 10]
```diff
@@ -211,6 +211,8 @@ struct ValidationInfo {
 
 struct FunctionValidator : public WalkerPass<PostWalker<FunctionValidator>> {
   bool isFunctionParallel() override { return true; }
+  // TODO: make this meaningfully true.
+  bool acceptsStackIR() override { return true; }
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/wasm/wasm-validator.cpp:215]: Yep, pretty much.

File: src/wasm2js.h [hunk 11]
```diff
@@ -410,9 +410,12 @@ Ref Wasm2JSBuilder::processWasm(Module* wasm, Name funcName) {
       ValueBuilder::makeName("// EMSCRIPTEN_START_FUNCS\n"));
   }
   // functions
-  ModuleUtils::iterDefinedFunctions(*wasm, [&](Function* func) {
-    asmFunc[3]->push_back(processFunction(wasm, func));
-  });
+  // Don't use iterDefinedFunctions because the iterator could be invalidated
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/wasm2js.h:413]: One of the functions called like five levels below this actually adds new utility functions to the module. This never caused problems before because previous optimizations removed functions, making room in the vector for the utility functions to be added without reallocating. I found this problem because a round trip I had here replaced the old vector with extra capacity with a smaller fresh vector.

File: scripts/fuzz_opt.py [hunk 12]
```diff
@@ -655,7 +655,16 @@ def write_commands(commands, filename):
...
+    ["-O0", "--roundtrip"],
+    ["-O1", "--roundtrip"],
+    ["-O2", "--roundtrip"],
+    ["-O3", "--roundtrip"],
+    ["-O4", "--roundtrip"],
+    ["-Os", "--roundtrip"],
+    ["-Oz", "--roundtrip"],
```
Comment (kripken) [inline scripts/fuzz_opt.py:667]: this worries me

File: src/ir/module-utils.h [hunk 13]
```diff
@@ -259,6 +259,9 @@ template<typename T> struct ParallelFunctionAnalysis {
     struct Mapper : public WalkerPass<PostWalker<Mapper>> {
       bool isFunctionParallel() override { return true; }
       bool modifiesBinaryenIR() override { return false; }
+      // Trust that the caller will behave correctly
```
Comment (kripken) [inline src/ir/module-utils.h:262]: please expand the comment a little more, specifically, what is "correct" here.

File: src/pass.h [hunk 14]
```diff
@@ -180,7 +180,7 @@ struct PassRunner {
 
   // Adds the default set of optimization passes; this is
   // what -O does.
-  void addDefaultOptimizationPasses();
+  void addDefaultOptimizationPasses(bool withStackIR = true);
```
Comment (kripken) [inline src/pass.h:183]: In general we've avoided such APIs. How about something like
```
enum StackIRUsage { NoStackIR, WithStackIR };
void addDefaultOptimizationPasses(StackIRUsage stackIRUsage = WithStackIR)
```
then the caller would be `addDefaultOptimizationPasses(NoStackIR)` which seems clearer

File: src/pass.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/pass.h:300]: I think StackIR/BinaryenIR are the terms we've used so far. Makes sense to me to continue with those.

File: src/pass.h (hunk 14 above)
Comment (kripken) [reply src/pass.h:183]: (but see my main comment, which might avoid the need for this part of the change I think)

**Comment by tlively** [comment]
> The current approach seems like it breaks `wasm-opt -O3 -O3` or `wasm-opt -O3 --converge`. That indicates to me that we should look at this differently. How about if instead of making `-O3` add stack IR as part of the list of passes, we don't do it there. Instead, the entire `wasm-opt` invocation adds stack IR at the very end, or does not, depending on the optimization level. That is, `wasm-opt -O3 -O3` would mean "run `-O3` opts twice, and also set the optimization level to 3, which will do Stack IR while writing"
> 
> That's a fairly small change but it feels more correct to me: what wasm-opt does is convert to binaryen IR, do passes, then convert to wasm. Stack IR is really part of that last stage.

I see where you're coming from, but the theme of this change is making Stack IR a more first-class citizen, so I feel it would incongruous to also treat it as an extra add-on at the end in the same patch. For now I think `wasm-opt -O3 -O3` and friends should just be viewed as incorrect invocations to which users need to add roundtrips. This is not so different from the situation with the passes that require `--flatten` to be run beforehand. As a long term solution, my hope is that converting between stacky and non-stacky forms of the IR can be made fast enough that the pass runner would make the conversions automatically as necessary.

File: src/pass.h [hunk 15]
```diff
@@ -297,6 +291,14 @@ class Pass {
   // out any Stack IR - it would need to be regenerated and optimized.
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/pass.h:291]: Ping

File: src/pass.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/pass.h:300]: What I meant was, then why do have two different names (BinaryenIR and NestedIR) for the same thing?

File: src/passes/passes.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/passes.h:86]: So are you gonna remove it?

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/passes.h:86]: Yikes, I thought I had, but I did some very nasty merges and I guess I lost the change. I'll look everything over again to make sure it looks right.

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/passes.h]: Oh lol the issue was that I didn't push after my nasty merges.

**Comment by kripken** [comment]
> For now I think wasm-opt -O3 -O3 and friends should just be viewed as incorrect invocations to which users need to add roundtrips. 

This does really go to what Stack IR is - just part of writing the binary, or a full secondary IR. And how that should be surfaced to users - should `-O3 -O3` start to be an error, or not. These seem like big questions, so we should consider them carefully.

Let me try to briefly summarize them:

 * Stack IR is part of binary writing. We can optimize while doing so, or not. `-O3 -O3` is fine, we don't write a binary as part of that. As Stack IR is only emitted right before writing the binary, and immediately thrown out, it can never go stale or be emitted unnecessarily.
 * Stack IR is a full secondary IR. `-O3 -O3` fails because the first lowers to Stack IR which we can't run the whole optimization pipeline on. Running individual Stack IR passes is easy in this way as they are just more passes (and not part of binary writing), they are first-class citizens.

File: scripts/fuzz_opt.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tlively) [reply scripts/fuzz_opt.py:662]: > Like other passes, `-O0` or `-O1` do not end with stack IR because they don't contain `--generate-stack-ir`, no?

`-O0` and `-O1` *do* contain --generate-stack-ir when some later option like `-O3` overrides the shrink and opt levels.

File: src/pass.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/pass.h:291]: I guess if this is all `modifiesBinaryenIR` was used for, then we probably don't need that function at all any more. Can you think of anything we might use it for in the future?

File: src/pass.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/pass.h]: The PR originally had two different names because I couldn't decide whether I wanted to rename it or not so I only renamed it in half the relevant places. My bad 😅

File: src/pass.h (hunk 14 above)
Comment (tlively) [reply src/pass.h:183]: Sure, I can make this change but let's figure out the higher-level question first. Searches for "= true)" and "= false)" show a few more places where this pattern could be cleaned up as well.

**Comment by tlively** [comment]
I strongly prefer that Stack IR be a first-class IR for now and not viewed as just a part of binary writing because multivalue makes "stacky" optimizations more attractive and sometimes necessary, as in the case of cleaning up the lowered forms of Binaryen's tuple instructions.

However, as a follow up I would very much like to remove Stack IR and let Binaryen IR handle these stacky optimizations, as in #2784. That would let us remove much (but not all) of the complexity of having two IRs.

File: src/pass.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/pass.h]: Sorry, I somehow saw this part remaining in the previous version, but now it's gone, so it doesn't matter.

[the remark this answers is not in the document]
Comment (tlively) [reply src/pass.h]: No, you were right! I hadn't fixed the comment before :) Only when I went to go fix it earlier did I realize we didn't need `modifiesBinaryenIR` at all.

**Comment by aheejin** [comment]
@tlively, sorry, you explained to me the rationale behind why you want stack IR to be the first class like two weeks ago and I thought I understood it, but I don't remember details now 😅 Could you summarize that again?

And yeah, making `-O3 -O3` invalid looks a bit worrisome.

**Comment by tlively** [comment]
> @tlively, sorry, you explained to me the rationale behind why you want stack IR to be the first class like two weeks ago and I thought I understood it, but I don't remember details now 😅 Could you summarize that again?

The rationale is that we need to lower away tuple operations when we convert to Stack IR so that we can optimize the stacky multivalue code they produce. But that lowering introduces new locals on the Function object, so the Stack IR would no longer be able to be simply thrown out to restore the Function to its original condition. This PR "fixes" the problem by turning places where the code used to silently throw away Stack IR into errors.

**Comment by tlively** [comment]
Maybe this intermediate state with Stack IR as a somewhat-first-class IR is too half baked and complicated. Would it be better if I pushed on #2784 as an opt-in experimental feature until it was fullly mature so we could switch wholesale to a new, mature system?

File: scripts/fuzz_opt.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply scripts/fuzz_opt.py:662]: Oh, I didn't know that parsing rule...
